File: trab2/ruby/tabela.rb
require './pprint'

include Math
PIE = PI / E

module Enumerable
  def mapred; map{|i| @t = yield(i, @t)} end
  def myfold(init) mapred{|i, p| p.nil? ? init : yield(i, p)}.reduce(:+) end
end

def mysin(x) (0..@n).myfold(x){|k, z| -z * x * x / (2.0 * k) / (2.0 * k + 1.0)} end
def mycos(x) (0..@n).myfold(1){|k, z| -z * x * x / (2.0 * k) / (2.0 * k - 1.0)} end
def mytan(x) mysin(x) / mycos(x) end
def mypi; 4 * (0..@n).myfold(1){|k, z| -z * (2.0 * k - 1.0) / (2.0 * k + 1.0)} end
def mye; (0..@n).myfold(1.0){|k, z| z / k} end
def mypie; mypi / mye end

tabela = lambda do |n|
  @n = n
  # calcular valores
  valores =
    ["n=#{@n}", 'Ve', 'Va', 'Eabs', 'Erel'],
    ['pi', PI, mypi],
    ['e', E, mye],
    ['pi/e', PIE, mypie],
    ['sen(pi/e)', sin(PIE), mysin(mypie)],
    ['cos(pi/e)', cos(PIE), mycos(mypie)],
    ['tg(pi/e)', tan(PIE), mytan(mypie)]
  #calcular erros
  for v in valores[1..-1] do
    ve, va = v[1], v[2]
    v << eabs = ve - va
    v << erel = eabs / ve
  end
  # imprimir tabela e uma linha em branco
  print_table valores
  puts
end

[3, 10, 30, 84].map &tabela

File: trab2/erlang/tabela.erl
-module(tabela).
-export([main/0]).
-import(math, [pi/0, exp/1, sin/1, cos/1, tan/1]).
-import(lists, [seq/2, nth/2, append/2, map/2, foldl/3]).

myfold(F, I, N) -> {_, R} = foldl(fun(T, {Z, A}) -> {Z * F(T), A + Z} end, {I, 0}, seq(1, N)), R.

mysin(X, N) -> myfold(fun(K) -> -X * X / (2 * K) / (2 * K + 1) end, X, N).
mycos(X, N) -> myfold(fun(K) -> -X * X / (2 * K) / (2 * K - 1) end, 1, N).
mytan(X, N) -> mysin(X, N) / mycos(X, N).
mypi(N)     -> myfold(fun(K) -> -(2 * K - 1) / (2 * K + 1) end, 4, N).
mye(N)      -> myfold(fun(K) -> 1 / K end, 1, N).

tabela(N) ->
    PIE = pi() /exp(1),
    MYPIE = mypi(N) / mye(N),
    map(fun(ROW) ->
            Ve = nth(2, ROW),
            Va = nth(3, ROW),
            append(ROW, [Va - Ve, 1 - Ve / Va]) end,
    [["pi",        pi(),     mypi(N)        ],
     ["e",         exp(1),   mye(N)         ],
     ["pi/e",      PIE,      MYPIE          ],
     ["sin(pi/e)", sin(PIE), mysin(MYPIE, N)],
     ["cos(pi/e)", cos(PIE), mycos(MYPIE, N)],
     ["tg(pi/e)",  tan(PIE), mytan(MYPIE, N)]]).

main() ->
    map(fun(N) -> pprint:ptable([[N, "Ve", "Va", "Eabs", "Erel"]|tabela(N)]) end,
        [3, 10, 30, 84, 10000]).


